Testing book functionality

ShastkivRuslan opened the pull request.

ShastkivRuslan committed:
added dependencies;

ShastkivRuslan committed:
-added tests for BookRepository;

ShastkivRuslan committed:
-added tests for BookController;

ShastkivRuslan committed:
-minor;

ShastkivRuslan committed:
-minor;

ShastkivRuslan committed:
-final commit;

ShastkivRuslan committed:
-bug-fix;

ShastkivRuslan committed:
-build fix;

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,275 @@
+                .andExpect(MockMvcResultMatchers.status().isOk())
+                .andReturn();
+
+        String expectedString = objectMapper.writeValueAsString(expectedPage);
+        String actualString = result.getResponse().getContentAsString();
```
can you compare objects?

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,275 @@
+            Should throw EntityNotFoundException for an invalid book ID
+            """)
+    public void getBokById_BookByInvalidId_ThrowException() throws Exception {
+        System.out.println(this);
```
```suggestion
```

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,275 @@
+                .andExpect(MockMvcResultMatchers.status().isOk())
+                .andReturn();
+
+        String expectedString = objectMapper.writeValueAsString(expectedPage);
+        String actualString = result.getResponse().getContentAsString();
```
the same

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,275 @@
+        String actualString = result.getResponse().getContentAsString();
+
+        assertEquals(expectedString, actualString);
+
```
```suggestion
```

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,275 @@
+            Should return an empty page when no books match the search parameters
+            """)
+    public void search_InvalidSearchParams_ReturnEmptyPage() throws Exception {
+        MvcResult result = mockMvc.perform(
+                        MockMvcRequestBuilders.get(SEARCH_BOOK_URL)
+                                .param("titles", NON_EXISTING_TITLE)
+                                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(MockMvcResultMatchers.status().isOk())
+                .andReturn();
+        Page<BookDto> expectedPage = new PageImpl<>(List.of(), PAGEABLE, EMPTY_PAGE);
+        String expectedString = objectMapper.writeValueAsString(expectedPage);
+        String actualString = result.getResponse().getContentAsString();
+
+        assertEquals(expectedString, actualString);
```
Let’s  follow the Given-When-Then test structure

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/controller/CategoryControllerTest.java:
```diff
@@ -0,0 +1,208 @@
+},
+        executionPhase = Sql.ExecutionPhase.AFTER_TEST_CLASS)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+public class CategoryControllerTest {
```
Let’s apply the same comments ;)

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/controller/CategoryControllerTest.java:
```diff
@@ -0,0 +1,208 @@
+                result.getResponse().getContentAsString(), CategoryDto.class);
+
+        assertEquals(expected, actual);
+
```
```suggestion
```

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/repository/BookRepositoryTest.java:
```diff
@@ -0,0 +1,115 @@
+    void findAllWithCategories_getInitializedCategory_DoesNotThrowLazyInitializationException() {
+        List<Book> allWithCategories = bookRepository.findAllWithCategories(PAGEABLE).getContent();
+
+        for (Book book : allWithCategories) {
+            assertDoesNotThrow(() -> {
+                book.getCategories().size();
+            }, "LazyInitializationException was thrown for book: " + book.getTitle());
+        }
```
you can refactor it 
```
assertAll(
    allWithCategories.stream()
        .map(book -> () -> assertDoesNotThrow(
            () -> book.getCategories().size(),
            "LazyInitializationException was thrown for book: " + book.getTitle()
        ))
);
```

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/service/BookServiceTest.java:
```diff
@@ -0,0 +1,273 @@
+    private BookServiceImpl bookService;
+    @Mock
+    private BookRepository bookRepository;
+    @Mock
+    private BookMapper bookMapper;
```
Advanced (optional): you can use Spy and simplify your test cases

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/service/BookServiceTest.java:
```diff
@@ -0,0 +1,273 @@
+                -> bookService.findBookById(INVALID_BOOK_ID));
+    }
+}
+
```
```suggestion
```

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/service/CategoryServiceTest.java:
```diff
@@ -0,0 +1,169 @@
+        assertEquals(PAGEABLE.getPageSize(), actual.getSize());
+        assertEquals(List.of(firstCategoryDto, secondCategoryDto), actual.getContent());
+
+        // Verify interactions
```
```suggestion
```

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/service/CategoryServiceTest.java:
```diff
@@ -0,0 +1,169 @@
+        verify(categoryRepository).deleteById(FIRST_CATEGORY_ID);
+        verifyNoMoreInteractions(categoryRepository);
+    }
+
```
```suggestion
```

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/utils/BookTestUtils.java:
```diff
@@ -0,0 +1,113 @@
+import ruslan.shastkiv.bookstore.dto.book.CreateBookRequestDto;
+import ruslan.shastkiv.bookstore.model.Book;
+
+public class BookTestUtils {
```
🔥

liliia-ponomarenko commented on src/test/resources/scripts/book/insert_relations_between_book_and_categories.sql:
```diff
@@ -0,0 +1,4 @@
+INSERT INTO books_categories (book_id, category_id) VALUES
+    (1, 1),
+    (2, 2),
+    (3, 2);
```
add one empty line to the end of the file (check other files)

liliia-ponomarenko reviewed (changes requested):
Good job! Let’s  improve your solution ;)

ShastkivRuslan committed:
-changes after mentor review;

ShastkivRuslan commented:
Object comparison: Replaced string comparison of expectedString and actualString with proper object comparison in the test cases. Now the objects are compared directly, ensuring correctness and better readability.

Test structure improvement: Refactored tests to follow the Given-When-Then structure for better clarity and consistency in test case formatting.

Code refactoring: Simplified assertions and refactored repetitive code:

In BookRepositoryTest, replaced individual assertions with a more compact assertAll block to check for LazyInitializationException.
Simplified mock verifications and removed redundant interactions in tests.

Small formatting fixes: Added an empty line at the end of insert_relations_between_book_and_categories.sql as per standard project formatting practices.

sarakhmen reviewed (approved):
Great job!
